rembed/stack.py

```diff
@@ -146,6 +148,71 @@ def _make_inputs(self):
         self.X = self.X or T.imatrix("X")
         self.transitions = self.transitions or T.imatrix("transitions")
 
+    def _step(self, transitions_t, ss_mask_gen_matrix_t, stack_t, buffer_cur_t, 
+            tracking_hidden_prev, stack_pushed, stack_merged, buffer):
+        batch_size, _ = self.X.shape
+        # Extract top buffer values.
+        idxs = buffer_cur_t + (T.arange(batch_size) * self.seq_length)
+        buffer_top_t = buffer[idxs]
+
+        if self._predict_network is not None:
+            # We are predicting our own stack operations.
+            predict_inp = T.concatenate(
+                [stack_t[:, 0], stack_t[:, 1], buffer_top_t], axis=1)
+            if self.use_tracking_lstm:
+                # update the hidden state and obtain predicted actions
+                tracking_hidden, actions_t = self._predict_network(
+                tracking_hidden_prev, predict_inp, self.model_dim * 3, 
+                self.tracking_lstm_hidden_dim, self._vs, name="predict_actions")
+            else:
+                # obtain predicted actions
+                actions_t = self._predict_network(
+                predict_inp, self.model_dim * 3, util.NUM_TRANSITION_TYPES, self._vs,
+                name="predict_actions")
+                # no change in hidden state because there is none
+                tracking_hidden = tracking_hidden_prev
+        else:
+            # no change in the hidden state because there is none
+            tracking_hidden = tracking_hidden_prev
+
+        if self.use_predictions: 
+            # Predicting our own actions
+            if self.interpolate:
+                # Interpolate between truth and prediction using bernoulli RVs 
+                # generated prior to the step
+                mask = (transitions_t * ss_mask_gen_matrix_t + 
```

Comment:
Hi, it doesn't touch Models 1 or 2 - just Model12SS (I created a separate
option for schedule sampling). As of now experiments on SST and SNLI are
still running.

You're right about the operation being an elemwise multiplication to select
from the ground truth and the predictions.

Also, about the two step functions created - I kept running into errors on
generating the scheduled sampling mask from within the step function, and
having theano.scan to iterate over a None would raise a problem, so I let
this be for now. A cleaner alternative would be to pass a matrix of all 1s
instead of the RandomStreams-generated matrix, but I'm not sure if it would
not slow things down a teeny bit.

On Sun, Nov 29, 2015 at 8:22 PM, Jon Gauthier <notifications@github.com>
wrote:

> In rembed/stack.py
> <https://github.com/hans/rembed/pull/1#discussion_r46107907>:
>
> > +                # obtain predicted actions
> > +                actions_t = self._predict_network(
> > +                predict_inp, self.model_dim * 3, util.NUM_TRANSITION_TYPES, self._vs,
> > +                name="predict_actions")
> > +                # no change in hidden state because there is none
> > +                tracking_hidden = tracking_hidden_prev
> > +        else:
> > +            # no change in the hidden state because there is none
> > +            tracking_hidden = tracking_hidden_prev
> > +
> > +        if self.use_predictions:
> > +            # Predicting our own actions
> > +            if self.interpolate:
> > +                # Interpolate between truth and prediction using bernoulli RVs
> > +                # generated prior to the step
> > +                mask = (transitions_t * ss_mask_gen_matrix_t +
>
> (Maybe more for @raghavgupta93 <https://github.com/raghavgupta93> ..)
> Nice trick with ss_mask_gen_matrix_t. How does this affect Model 2
> performance? I expect not so much, since this is just an elemwise
> multiplication on a batch_size vector IIRC.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hans/rembed/pull/1/files#r46107907>.
>



-- 
Raghav Gupta
Graduate Student, Computer Science
Stanford University